design/apps.go

```diff
@@ -0,0 +1,291 @@
+	simpleDeploymentStatSeries,
+	nil)
+
+var simpleEnvironmentStatSingle = JSONSingle(
+	"EnvStats", "Holds a single response to a pipeline/stats request",
+	envStats,
+	nil)
+
+var _ = a.Resource("apps", func() {
+	a.BasePath("/apps")
+
+	// An auth token is required to call the auth API to get an OpenShift auth token.
+	a.Security("jwt")
+
+	a.Action("showSpace", func() {
+		a.Routing(
+			a.GET("/spaces/:spaceID"),
+		)
+		a.Description("list applications in a space")
+		a.Params(func() {
+			a.Param("spaceID", d.UUID, "ID of the space")
+		})
+		a.Response(d.OK, simpleSpaceSingle)
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("showSpaceApp", func() {
+		a.Routing(
+			a.GET("/spaces/:spaceID/applications/:appName"),
+		)
+		a.Description("list application")
+		a.Params(func() {
+			a.Param("spaceID", d.UUID, "ID of the space")
+			a.Param("appName", d.String, "Name of the application")
+		})
+		a.Response(d.OK, simpleAppSingle)
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("showSpaceAppDeployment", func() {
+		a.Routing(
+			a.GET("/spaces/:spaceID/applications/:appName/deployments/:deployName"),
+		)
+		a.Description("list pipe element")
+		a.Params(func() {
+			a.Param("spaceID", d.UUID, "ID of the space")
+			a.Param("appName", d.String, "Name of the application")
+			a.Param("deployName", d.String, "Name of the pipe deployment")
+		})
+		a.Response(d.OK, simpleDeploymentSingle)
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("showDeploymentStats", func() {
+		a.Routing(
+			a.GET("/spaces/:spaceID/applications/:appName/deployments/:deployName/stats"),
+		)
+		a.Description("get deployment statistics")
+		a.Params(func() {
+			a.Param("spaceID", d.UUID, "ID of the space")
+			a.Param("appName", d.String, "Name of the application")
+			a.Param("deployName", d.String, "Name of the deployment")
+			a.Param("start", d.Number, "start time in millis")
+		})
+		a.Response(d.OK, simpleDeploymentStatsSingle)
+		a.Response(d.Unauthorized, JSONAPIErrors)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
+	})
+
+	a.Action("showDeploymentStatSeries", func() {
+		a.Routing(
+			a.GET("/spaces/:spaceID/applications/:appName/deployments/:deployName/statseries"),
+		)
+		a.Description("list deployment statistics")
+		a.Params(func() {
+			a.Param("spaceID", d.UUID, "ID of the space")
+			a.Param("appName", d.String, "Name of the application")
+			a.Param("deployName", d.String, "Name of the deployment")
+			a.Param("start", d.Number, "start time in millis")
+			a.Param("end", d.Number, "end time in millis")
+			a.Param("limit", d.Integer, "maximum number of data points to return")
```

Comment:
I'm not sure this conforms to pagination being done in JSONAPI. I think we're doing that differently elsewhere (e.g. `search` endpoint). @aslakknutsen @xcoulon your thoughts?

Reply:
This is not really pagination, in that it represents a time frame with a variable number of return values.  The parameters are supposed to be easily passed on to Hawkular.